### ISSUE TITLE
e2e test scenario for extension

### ISSUE THREAD

**Issue by DonJayamanne** [microsoft/vscode-jupyter#9584]
### Discussed in https://github.com/microsoft/vscode-jupyter/discussions/9575

<div type='discussions-op-text'>

<sup>Originally posted by **christian-bromann** April  4, 2022</sup>
Hey extension devs 👋 

I've been working on an [e2e test integration](https://github.com/webdriverio-community/wdio-vscode-service) into WebdriverIO, a test framework based on the W3C WebDriver standard. At my company we needed e2e testing for some of our extensions and I wonder if the you would be interested to add this to the Jupyter VSCode Extension as well. Tests can be run using GH Actions on Mac, Windows or Ubuntu. It seems that the project already has tons of unit tests and I am sure there are some e2e integration scenarios we could cover with e2e testing.

I am happy to take a stab at this if this would be something y'all would consider.</div>

**Comment by christian-bromann** [comment]
To give an outlook how these tests would look like:

- in the [service integration test suite](https://github.com/webdriverio-community/wdio-vscode-service/blob/main/test/specs/basic.e2e.ts) I am currently building the test suite for the page objects so it shows how the interface with VSCode components look like, the TypeScript interface is documented in the [project docs](https://webdriverio-community.github.io/wdio-vscode-service/)
- In [`vscode-marquee`](https://github.com/stateful/vscode-marquee/blob/main/test/specs/test.e2e.ts) we are already using this integration to test our webviews of the extension leveraging the proposed page object design

Let me know if there are any questions / concerns.

**Comment by DonJayamanne** [comment]
Thanks, we'll discuss this in our next triage meeting:
Internal notes for the team:
* At one point we had such tests (BDD tests using puppeteer) in the Python extension
	* Pros
		* Creating BDD tests was easy (but the BDD tests were never used by anyone, the plan was PMs and others would write tests that were targetting golden scenarios or the like)
		* I.e. BDD was the only benefit
	* Cons
		* When ever VSC UI changes, the CSS queries have to be updated everytime with ever release, thus we'd end up with one set of tests working for Stable & not insiders (unless the CSS queries are updated)
			* Looks like this is still an issue https://github.com/webdriverio-community/wdio-vscode-service/blob/main/src/locators/1.44.0.ts
			* I.e. tests will not run against insiders until this package is updated.
			* When I built the same tests for Python extension, engineers from vscode approached me about this same issue and how I circumvented those issues (it wasn't an issue as we were always targetting stable & I had time to update our scripts).
		* Slow (we need to wait for the UI to get updated, to be visible).
			* E.g. if we want to run a cell, then the cell must be visible in the notebook UI to be able to click the run button, where as with the VS Code API we don't need to wait for cell to be visible.
			* This alone could be a major blocker.
			* This was an issue in Interactive Window tests (the BDD tests I wrote), and I had to scroll the IW window, similarly an issue with IW Python file, I had to scroll the python text file so that the code comes into view and code lenses are generated and visible. Where as with the API we can just send text to the IW without code lenses being visible.
	* TLDR - The BDD tests using puppeteer was deleted completely

**Comment by christian-bromann** [comment]
> * Looks like this is still an issue

Not sure why though. As you can see we are maintaining locators for every version of VSCode to ensure that the page object interfaces continue to work after VSCode updates.

> * I.e. tests will not run against insiders until this package is updated.

I am currently working on setting up regular tests against insiders and the last 3 stable versions of VSCode. Any recommendation here how many versions back it is important to test?

> * This alone could be a major blocker.

I haven't yet set up page objects for Notebooks but I assume here that VSCode only renders these cells that are within the viewport which indeed can slow down the tests as some sort of iterative scrolling behavior needs to be imitated. However it is certainly possible to build a bridge to the VSCode API and circumvent this issue. At the end of the day the UI doesn't always need to be involved in the test.

**Comment by greazer** [comment]
Thanks for the feedback! However, we don't have plans on making any changes to how we do our automated testing at this time.

**Comment by christian-bromann** [comment]
@greazer I was not suggesting to make any changes to existing test, just to add specific tests that would cover certain e2e scenarios, which aren't covered with the current setup. I am happy to take a stab at this if your team likes to add more coverage.
